wire/msgtx_test.go

```diff
@@ -778,6 +779,149 @@ func TestTxWitnessSize(t *testing.T) {
 	}
 }
 
+// TestTxWitnessLimits performs negative tests to ensure decoding
+// serialized transactions with witness data triggers proper error paths.
+func TestTxWitnessLimits(t *testing.T) {
```

Comment:
thinking more about it... if these limit checks are already happening deeper in consensus logic, I wonder if it doesn't make more sense to simply forget about witness limit checks and just do a <4MB size check on the entire txn? if these checks at a parsing level are really more about memory exhaustion safeguards, there shouldn't be a need to look for the witness specifically?

Reply:
> , I wonder if it doesn't make more sense to simply forget about witness limit checks and just do a <4MB size check on the entire txn?

That would certainly be the most conservative check. The issue with doing it on this layer though, is the possibility of future interactions that increase this value (say an ext block in the future or w/e). On the bitcoind side, they have such limits in place still, but it's around 32 MB or so (iirc the max p2p payload). So if we end up upgrading Bitcoin far in the future to 64 MB blocks or w/e, then that would cause a fork with these old clients.